src/Lean/Meta/Tactic/Simp/Main.lean

```diff
@@ -571,10 +571,127 @@ def congr (e : Expr) : SimpM Result := do
   else
     congrDefault e
 
+/--
+Returns `true` if `e` is of the form `@letFun _ (fun _ => β) _ _`
+
+`β` must not contain loose bound variables. Recall that `simp` does not have support for `let_fun`s
+where resulting type depends on the `let`-value. Example:
+```
+let_fun n := 10;
+BitVec.zero n
+```
+-/
+def isNonDepLetFun (e : Expr) : Bool :=
+  let_expr letFun _ beta _ body := e | false
+  beta.isLambda && !beta.bindingBody!.hasLooseBVars && body.isLambda
```

Comment:
I thought I'd mention that `Lean.Expr.letFun?` is an existing recognizer for letfuns, though it doesn't give anything in this case (it's able to deal with letfuns that have a non-lambda body, but that is a dependent letfun). There's also `Lean.Expr.letFunAppArgs?` to handle the case where the letfun is applied as a function (like `(let_fun f := id; f) 2`), which is used elsewhere to handle reducing letfuns.